code/HabitTracker/app/src/main/java/com/example/habittracker/utils/CustomHabitList.java

```diff
@@ -51,11 +62,113 @@ public View getView(int position, @Nullable View convertView, @NonNull ViewGroup
         // get the views
         TextView habitTitle = view.findViewById(R.id.habitTitle);
         TextView habitReason = view.findViewById(R.id.habitReason);
+        ProgressBar progressBar = view.findViewById(R.id.progressBar);
+        TextView progressText = view.findViewById(R.id.progressValue);
+
+        final Button moveUp = view.findViewById(R.id.move_habit_up);
+        final Button moveDown = view.findViewById(R.id.move_habit_down);
+
+        if(!keepOn){
+            moveUp.setVisibility(View.GONE);
+            moveDown.setVisibility(View.GONE);
+        }
+
+        if(!buttonList.contains(moveUp)){
+            buttonList.add(moveUp);
+        }
+        if(!buttonList.contains(moveDown)){
+            buttonList.add(moveDown);
+        }
+
+        moveUp.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                if(position>0){
+                    Habit h1 = habitList.get(position);
+                    Habit h2 = habitList.get(position-1);
+                    habitList.set(position-1,h1);
+                    habitList.set(position,h2);
+                    HashMap<String,Object> doc1 = h1.toDocument();
+                    doc1.put("order",position-1);
+                    HashMap<String,Object> doc2 = h2.toDocument();
+                    doc2.put("order",position);
+                    DatabaseManager.get().updateHabitDocument(SharedInfo.getInstance().getCurrentUser().getUsername(),h1.getTitle(),h1.getTitle(),doc1);
+                    DatabaseManager.get().updateHabitDocument(SharedInfo.getInstance().getCurrentUser().getUsername(),h2.getTitle(),h2.getTitle(),doc2 );
+                }
+                notifyDataSetChanged();
+            }
+        });
+
+        moveDown.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                if(position<habitList.size()-1){
+                    Habit h1 = habitList.get(position);
+                    Habit h2 = habitList.get(position+1);
+                    habitList.set(position+1,h1);
+                    habitList.set(position,h2);
+                    HashMap<String,Object> doc1 = h1.toDocument();
+                    doc1.put("order",position+1);
+                    HashMap<String,Object> doc2 = h2.toDocument();
+                    doc2.put("order",position);
+                    DatabaseManager.get().updateHabitDocument(SharedInfo.getInstance().getCurrentUser().getUsername(),h1.getTitle(),h1.getTitle(),doc1);
+                    DatabaseManager.get().updateHabitDocument(SharedInfo.getInstance().getCurrentUser().getUsername(),h2.getTitle(),h2.getTitle(),doc2 );
+                }
+                notifyDataSetChanged();
+            }
+        });
+
+//        if(reorder != null){
```

Comment:
delete this maybe, if not going to use it

Reply:
Deleted